oak/server/rust/oak_runtime/tests/integration_test.rs

```diff
@@ -0,0 +1,116 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use std::thread::{park, spawn};
+
+mod common {
+    use oak_runtime::runtime::{Handle, Runtime};
+
+    use hyper::{Client, Uri};
+    use log::info;
+    use maplit::hashmap;
+    use oak_abi::OakStatus;
+    use oak_runtime::{config, metrics};
+    use std::sync::Arc;
+    use wat::parse_str;
+
+    const PORT: u16 = 9876;
+
+    pub fn start_runtime() -> Result<(Arc<Runtime>, Handle), OakStatus> {
+        let wat = r#"
+        (module
+            (type (;0;) (func (param i64) (result i32)))
+            (func $oak_main (type 0)
+              i32.const 42)
+            (memory (;0;) 18)
+            (export "memory" (memory 0))
+            (export "oak_main" (func $oak_main)))
+        "#;
+        let binary = parse_str(wat).expect("Could not parse wat module.");
+
+        // Create a runtime with one node
+        let cfg = oak_runtime::application_configuration(
+            hashmap![
+                "node".to_string() => binary,
+            ],
+            "lumberjack",
+            "node",
+            "main",
+        );
+
+        info!("Starting the runtime with one nodes.");
+        config::configure_and_run(cfg)
+    }
+
+    pub fn start_metrics_server() {
+        info!("Starting the metrics server");
+
+        let mut tokio_runtime =
+            tokio::runtime::Runtime::new().expect("Couldn't create Tokio runtime");
+        let result = tokio_runtime.block_on(metrics::serve_metrics(PORT));
+
+        info!("Exiting metrics server node thread {:?}", result);
+    }
+
+    pub async fn read_metrics() -> Result<String, hyper::error::Error> {
+        info!("Reading the metrics.");
+        let client = Client::new();
+
+        let uri = format!("http://localhost:{}", &PORT)
+            .parse::<Uri>()
+            .expect("Could not parse URI.");
+        match client.get(uri).await {
+            Err(e) => Err(e),
+            Ok(res) => {
+                info!("status: {}", res.status());
+
+                // Concatenate the body stream into a single buffer...
+                match hyper::body::to_bytes(res).await {
+                    Err(e) => Err(e),
+                    Ok(buf) => Ok(format!("{:?}", buf)),
+                }
+            }
+        }
+    }
+}
+
+#[test]
+fn test_metrics_gives_the_correct_number_of_nodes() {
```

Comment:
@daviddrysdale Is this anything like the e2e test you described the other day?

Reply:
Yep, definitely heading that way &ndash; nice!  

However, these kind of tests do end up being quite brittle &ndash; small changes to the code give test breakages that aren't obviously correlated with the change.  

It would be nice to reduce the brittleness as much as possible, to make life easier for developers. This initial test should be fairly robust, because there's not much that can change with a single Node that immediately exits.  However, checking the exact help text makes the test a bit of a change-detector &ndash; it's probably worth adding a helper that extracts a particular numeric value from the result of `read_metrics`, so that (say) changes to the help text for a metric don't result in lots of broken tests.

Maybe a `read_int_metric(metric_name: &str) -> Option<i64>` method that just pulls a specific value?  (I'd hope that there's a Prometheus client library somewhere that can help with this).

Longer term, it might also be nice to have a test that's even more end-to-end: we could (say) change the `abitest` client binary to query the metrics (externally, over HTTP) at start-of-day, and check that the expected Node count is there (4 at the moment).  That definitely would be a brittle test though &ndash; if you ran the client twice in a row without restarting the Runtime, the count would be out &ndash; so we'd probably need to think about something that's more robust.